Function is_normal for covered and affine schemes

paemurru opened the pull request:
Also fixed a bug in `is_normal` for quotient rings: now it checks that the ring is reduced. This bug is still present in Singular ( https://www.singular.uni-kl.de/hannes/sing_1485.htm#SEC1566 ).

paemurru committed:
Function is_normal for covered and affine schemes

Also fixed a bug in `is_normal` for quotient rings: now it checks that the ring is reduced. This bug is still present in Singular.
https://www.singular.uni-kl.de/hannes/sing_1485.htm#SEC1566

paemurru committed:
Fix doctest

afkafkafk13 commented on src/Rings/mpoly-affine-algebras.jl:
```diff
@@ -931,25 +941,29 @@ julia> is_normal(A)
+    !check || is_reduced(A) || return false
+
+    I = A.I
+    # TODO remove old1 & old2 once new Singular jll is out
+    # QUESTION: How old is this? Has this been fixed yet?
```
@hannes14 : Do you have an idea whether this has already propagated up or not?
                         This would be the moment to get rid of the workaround, if it is not needed any more.

afkafkafk13 commented on test/Rings/mpoly_affine_algebras.jl:
```diff
@@ -37,8 +37,9 @@
-  @test is_normal(quo(R, ideal(R, [x - y^3]))[1])
+  @test !is_normal(quo(R, ideal(R, [x^2 - y^3]))[1]; check=false)
+  @test is_normal(quo(R, ideal(R, [x - y^3]))[1]; check=false)
+  @test !is_normal(quo(R, ideal(R, [x^2]))[1])
```
Please add [z^2-x*y] to the test cases.

paemurru replied:
Added.

afkafkafk13 commented:
This looks fine. The only reason, why I have not approved yet, is the question about the old bugfix.

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Methods.jl:
````diff
@@ -175,6 +175,31 @@ end
+julia> X = spec(R);
+
+julia> is_normal(X)
+```
````
A line of output seems to be missing here.

paemurru replied:
Fixed.

paemurru committed:
Fix doctest

paemurru committed:
Add test with z^2 - x*y

paemurru committed:
Merge branch 'oscar-system:master' into ep/is_normal

paemurru committed:
Propagate keyword argument `check`

Now, if `check` is set to `false` when testing a covered scheme, it is not going to test reducedness when calling `is_normal` for the underlying quotient rings.

afkafkafk13 reviewed (approved):
Thank you. The code looks good.

hannes14 commented:
Last important change top normal.lib was  Wed Nov 8 2023 (and has nothing to do with options).
What was the bug this work-around was for?
There were fixes for handling computation in quotient rings, but this seems to be something different.

afkafkafk13 commented:
> Last important change top normal.lib was Wed Nov 8 2023 (and has nothing to do with options). What was the bug this work-around was for? There were fixes for handling computation in quotient rings, but this seems to be something different.

I am not even sure that it was a bug in the first place. It looks rather like a performance hack to make the Singular computation faster by not requiring reduction. What struck me as odd were the comments. Leaving this bit of code in the function is not a problem per se, but leaving it hanging around with a TODO-comment and a question was rather unclear. Either the hack can be removed or the comments should be removed to make this part of the code look less experimental (in the stable part of the tree!).

benlorenz commented:
I think this was done here [`f290fec` (#1270)](https://github.com/oscar-system/Oscar.jl/pull/1270/commits/f290feca4473bb30a8f09e475dbf3beea1c06905) to make sure that calling this singular function doesn't change the output for code that runs later. But it seems that was fixed in Singular in https://github.com/Singular/Singular/commit/74622283ad46295f0602f84dc752ccf62f699461.

afkafkafk13 commented:
> I think this was done here [`f290fec` (#1270)](https://github.com/oscar-system/Oscar.jl/pull/1270/commits/f290feca4473bb30a8f09e475dbf3beea1c06905) to make sure that calling this singular function doesn't change the output for code that runs later. But it seems that was fixed in Singular in [Singular/Singular@7462228](https://github.com/Singular/Singular/commit/74622283ad46295f0602f84dc752ccf62f699461).

Indeed! This is the history which explains the mysterious comments -- and obsoletes them (along with the option-setting code).  Thank you.
